### PR TITLE
Automatic buffering and preemptive flushing

### PR DESCRIPTION
**Pull request by kbogtob**
### What does this PR do?

**Introducing new breaking changes:**
It refactors the batch class into the MessageBuffer class. It removes the usage of the `#batch` method on the Statsd class as buffering is now automatic and always used. It flushes preemptively when reaching a given message count (`max_buffer_pool_size`) or 95% of the maximum buffer size (`max_buffer_payload_size`). If the user wants to flush data, they could call a new method `#flush` on Statsd. (that method is used a lot in the specs)

Adding new specific integration tests relevant to buffering to test edge cases.

### Motivation

Preparation of version 5 road map with automatic buffering, refactored telemetry (buffered as other messages) and async IO in a separate thread.

### Notes

We cannot use the optimized buffer size for UDP (1432 bytes) for now because of how telemetry is implemented. It eats part of the buffer size with a worst estimation of the size it could take. But in UDP, with a 1432 byte size, only a few global tags would make the telemetry allocate a size bigger of the buffer than its own size!

I removed this problem for now by increasing the default size of the buffer. I will add another PR after this one is merged to refactor telemetry as a "client" of the buffer instead of it allocating buffer size and using directly the low level Connection class.

### CODE REVIEW

File: lib/datadog/statsd.rb [hunk 1]
```diff
@@ -110,8 +122,23 @@ def initialize(
 
       @sample_rate = sample_rate
 
-      # we reduce max_buffer_bytes by a the rough estimate of the telemetry payload
-      @batch = Batch.new(connection, (max_buffer_bytes - telemetry.estimate_max_size))
+      max_buffer_payload_size ||= (transport_type == :udp ? UDP_DEFAULT_BUFFER_SIZE : UDS_DEFAULT_BUFFER_SIZE)
+
+      if max_buffer_payload_size <= 0
+        raise ArgumentError, 'max_buffer_payload_size cannot be < 0'
```
Comment (remeh) [inline lib/datadog/statsd.rb:128]: Nit: `<=`

File: lib/datadog/statsd.rb [hunk 2]
```diff
@@ -110,8 +122,23 @@ def initialize(
...
+      end
+
+      unless disable_telemetry
+        max_buffer_payload_size = max_buffer_payload_size - telemetry.estimate_max_size
+        raise ArgumentError, 'max_buffer_payload_size is not high enough to use telemetry' if max_buffer_payload_size <= 0
+      end
+
+      # we reduce max_buffer_payload_size by a the rough estimate of the telemetry payload
```
Comment (remeh) [inline lib/datadog/statsd.rb:136]: Nit: this comment should be line 130

File: lib/datadog/statsd/message_buffer.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        @buffer = String.new
+        @message_count = 0
+
+        @depth = 0
+      end
+
+      def add(message)
+        message_size = message.bytesize
+
```
Comment (remeh) [inline lib/datadog/statsd/message_buffer.rb:29]: I'm wondering if we should add a check on the message_size == 0, to avoid adding useless `\n` to the buffer.

File: lib/datadog/statsd/message_buffer.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module Datadog
+  class Statsd
+    class MessageBuffer
+      PAYLOAD_SIZE_TOLERANCE = 0.05
```
[the remark this answers is not in the document]
Comment (kbogtob) [reply lib/datadog/statsd/message_buffer.rb:6]: My idea was to send messages preemptively when we *think* that we can't send a new message. You're right that when using a size of 1432, it's quite a lot. Do you think I should remove it?

File: lib/datadog/statsd/message_buffer.rb (hunk 3 above)
Comment (kbogtob) [reply lib/datadog/statsd/message_buffer.rb:27]: I think it's a good idea.

**Comment by iancanderson** [comment]
@kbogtob I'm interesting in the performance gains from this and https://github.com/DataDog/dogstatsd-ruby/pull/151

Are there any plans to push a new gem release with these PRs? Thanks!

File: lib/datadog/statsd.rb [hunk 5]
```diff
@@ -51,67 +57,59 @@ def tags
...
       port = nil,
+      socket_path: nil,
+
       namespace: nil,
       tags: nil,
-      max_buffer_bytes: DEFAULT_BUFFER_SIZE,
-      socket_path: nil,
-      logger: nil,
       sample_rate: nil,
-      disable_telemetry: false,
```
Comment (wjessop) [inline lib/datadog/statsd.rb:83]: It looks like this was a breaking API change that didn't trigger a changelog entry or major version bump. I suggest at least retro-actively adding a changelog entry, I had to dig around in the source to find this out.

Comment (remeh) [reply lib/datadog/statsd.rb:83]: Hey @wjessop thanks for your comment and sorry for the inconvenience.

I'll add a note about this in the 4th bullet point of the changelog entry of the 5.0.0 major version [here](https://github.com/DataDog/dogstatsd-ruby/blob/master/CHANGELOG.md#500--20210407), that the telemetry parameter has been renamed as well 👍

Comment (remeh) [reply lib/datadog/statsd.rb:83]: It'll go in with: https://github.com/DataDog/dogstatsd-ruby/pull/288
